cluster/v1alpha1/helpers.go

```diff
@@ -110,90 +116,166 @@ func (r *RolloutHandler) getRolloutAllClusters(rolloutStrategy RolloutStrategy,
+
+	// Check for removed Clusters
+	currentClusterStatus, removedClusterStatus := r.getRemovedClusters(allClusterGroups, existingClusterStatus)
+	rolloutResult := progressivePerCluster(allClusterGroups, len(allClusters), failureTimeout, currentClusterStatus)
+	rolloutResult.ClustersRemoved = removedClusterStatus
 
 	return &strategy, rolloutResult, nil
 }
 
-func (r *RolloutHandler) getProgressiveClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+func (r *RolloutHandler[T]) getProgressiveClusters(rolloutStrategy RolloutStrategy, existingClusterStatus []ClusterRolloutStatus) (*RolloutStrategy, RolloutResult, error) {
 	// Prepare the rollout strategy
 	strategy := RolloutStrategy{Type: Progressive}
 	strategy.Progressive = rolloutStrategy.Progressive.DeepCopy()
 	if strategy.Progressive == nil {
 		strategy.Progressive = &RolloutProgressive{}
 	}
 
-	// Upgrade mandatory decision groups first
-	groupKeys := decisionGroupsToGroupKeys(strategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
-	clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys...)
-
-	// Perform progressive rollout for mandatory decision groups
-	rolloutResult := progressivePerGroup(clusterGroups, maxTimeDuration, statusFunc)
-	if len(rolloutResult.ClustersToRollout) > 0 {
-		return &strategy, rolloutResult, nil
-	}
-
 	// Parse timeout for non-mandatory decision groups
 	failureTimeout, err := parseTimeout(strategy.Progressive.Timeout.Timeout)
 	if err != nil {
 		return &strategy, RolloutResult{}, err
 	}
 
-	// Calculate the length for progressive rollout
-	totalClusters := r.pdTracker.ExistingClusterGroupsBesides().GetClusters()
-	length, err := calculateLength(strategy.Progressive.MaxConcurrency, len(totalClusters))
+	// Check for removed clusters
+	clusterGroups := r.pdTracker.ExistingClusterGroupsBesides()
+	currentClusterStatus, removedClusterStatus := r.getRemovedClusters(clusterGroups, existingClusterStatus)
+
+	// Upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroups = r.pdTracker.ExistingClusterGroups(groupKeys...)
+
+	// Perform progressive rollOut for mandatory decision groups first.
+	if len(clusterGroups) > 0 {
+		rolloutResult := progressivePerGroup(clusterGroups, failureTimeout, currentClusterStatus)
+		if len(rolloutResult.ClustersToRollout) > 0 || len(rolloutResult.ClustersTimeOut) > 0 {
+			rolloutResult.ClustersRemoved = removedClusterStatus
+			return &strategy, rolloutResult, nil
+		}
+	}
+
+	// Calculate the length for progressive rollOut
+	// If the MaxConcurrency not defined, total clusters length is considered as maxConcurrency.
+	clusterGroups = r.pdTracker.ExistingClusterGroupsBesides(groupKeys...)
+	length, err := calculateLength(strategy.Progressive.MaxConcurrency, len(clusterGroups.GetClusters()))
```

Comment:
Aside: This is outside of this PR's intent, but could we consider renaming `calculateLength()` to something more descriptive like `calculateRolloutSize()`?